NPE when unsubscribe in proxy mode.

## Stack trace
```
21:42:48.204 [pulsar-io-76-2:org.apache.pulsar.broker.service.nonpersistent.NonPersistentSubscription@391] ERROR org.apache.pulsar.broker.service.nonpersistent.NonPersistentSubscription - [non-persistent://public/default/a%2Fb][7f000001d49c61d44ed8] Error deleting subscription
java.util.concurrent.CompletionException: java.lang.NullPointerException
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273) ~[?:1.8.0_291]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280) [?:1.8.0_291]
	at java.util.concurrent.CompletableFuture$AsyncRun.run$$$capture(CompletableFuture.java:1643) [?:1.8.0_291]
	at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java) [?:1.8.0_291]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute$$$capture(AbstractEventExecutor.java:164) [netty-common-4.1.67.Final.jar:4.1.67.Final]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java) [netty-common-4.1.67.Final.jar:4.1.67.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469) [netty-common-4.1.67.Final.jar:4.1.67.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500) [netty-transport-4.1.68.Final.jar:4.1.68.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986) [netty-common-4.1.67.Final.jar:4.1.67.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.67.Final.jar:4.1.67.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.67.Final.jar:4.1.67.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_291]
Caused by: java.lang.NullPointerException
	at org.apache.pulsar.broker.service.nonpersistent.NonPersistentTopic.lambda$unsubscribe$48(NonPersistentTopic.java:1006) ~[pulsar-broker-2.9.0-rc-202110221101.jar:2.9.0-rc-202110221101]
	at java.util.concurrent.CompletableFuture$AsyncRun.run$$$capture(CompletableFuture.java:1640) ~[?:1.8.0_291]
	... 9 more
```

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/Connection.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/Connection.java
@@ -29,19 +29,13 @@ import io.streamnative.pulsar.handlers.mqtt.messages.factory.MqttDisConnAckMessa
 import io.streamnative.pulsar.handlers.mqtt.support.handler.AckHandler;
 import io.streamnative.pulsar.handlers.mqtt.support.handler.AckHandlerFactory;
 import io.streamnative.pulsar.handlers.mqtt.utils.MqttUtils;
-import io.streamnative.pulsar.handlers.mqtt.utils.NettyUtils;
 import io.streamnative.pulsar.handlers.mqtt.utils.WillMessage;
-import java.util.Map;
 import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
 import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
 import lombok.Getter;
 import lombok.extern.slf4j.Slf4j;
-import org.apache.commons.lang3.tuple.Pair;
-import org.apache.pulsar.broker.service.BrokerServiceException;
-import org.apache.pulsar.broker.service.Consumer;
-import org.apache.pulsar.broker.service.Subscription;
-import org.apache.pulsar.broker.service.Topic;
 
 /**
  * Value object to maintain the information of single connection, like ClientID, Channel, and clean
@@ -74,6 +68,8 @@ public class Connection {
     @Getter
     private final MQTTConnectionManager manager;
     @Getter
+    private final TopicSubscriptionManager topicSubscriptionManager;
+    @Getter
     private final MqttConnectMessage connectMessage;
     @Getter
     private final AckHandler ackHandler;
@@ -105,6 +101,7 @@ public class Connection {
         this.keepAliveTime = builder.keepAliveTime;
         this.ackHandler = AckHandlerFactory.of(protocolVersion).getAckHandler();
         this.channel.attr(ATTR_KEY_CONNECTION).set(this);
+        this.topicSubscriptionManager = new TopicSubscriptionManager();
         this.addIdleStateHandler();
     }
 
@@ -138,59 +135,11 @@ public class Connection {
         return channel.close();
     }
 
-    public void removeConsumers() {
-        Map<Topic, Pair<Subscription, Consumer>> topicSubscriptions = NettyUtils
-                .getTopicSubscriptions(channel);
-        // For producer doesn't bind subscriptions
-        if (topicSubscriptions != null) {
-            topicSubscriptions.forEach((k, v) -> {
-                try {
-                    v.getLeft().removeConsumer(v.getRight());
-                } catch (BrokerServiceException ex) {
-                    log.warn("subscription [{}] remove consumer {} error", v.getLeft(), v.getRight(), ex);
-                }
-            });
-        }
+    public CompletableFuture<Void> close() {
+        return close(false);
     }
 
-    private void doRemoveSubscriptions() {
-        Map<Topic, Pair<Subscription, Consumer>> topicSubscriptions = NettyUtils
-                .getTopicSubscriptions(channel);
-        // For producer doesn't bind subscriptions
-        if (topicSubscriptions != null) {
-            topicSubscriptions.forEach((k, v) -> {
-                k.unsubscribe(this.clientId);
-                v.getLeft().delete();
-            });
-        }
-    }
-
-    public void removeSubscriptions() {
-        removeConsumers();
-        if (cleanSession) {
-            doRemoveSubscriptions();
-            // when mqtt client support session expire interval variable
-        } else {
-            // when use mqtt5.0 we need to use session expire interval to remove session.
-            // but mqtt protocol version lower than 5.0 we don't do that.
-            if (MqttUtils.isMqtt5(protocolVersion)
-                    && SESSION_EXPIRE_INTERVAL_UPDATER.get(this)
-                    != SessionExpireInterval.NEVER_EXPIRE.getSecondTime()) {
-                if (sessionExpireInterval == SessionExpireInterval.EXPIRE_IMMEDIATELY.getSecondTime()) {
-                    doRemoveSubscriptions();
-                } else {
-                    manager.newSessionExpireInterval(__ ->
-                            doRemoveSubscriptions(), clientId, SESSION_EXPIRE_INTERVAL_UPDATER.get(this));
-                }
-            }
-        }
-    }
-
-    public void close() {
-        close(false);
-    }
-
-    public void close(boolean force) {
+    public CompletableFuture<Void> close(boolean force) {
         if (log.isInfoEnabled()) {
             log.info("Closing connection. clientId = {}.", clientId);
         }
@@ -204,6 +153,24 @@ public class Connection {
             channel.writeAndFlush(mqttDisconnectionAckMessage);
         }
         this.channel.close();
+        // clean subscriptions
+        topicSubscriptionManager.removeAllConsumer();
+        if (cleanSession) {
+            return topicSubscriptionManager.removeAllSubscriptions();
+        }
+        // when use mqtt5.0 we need to use session expire interval to remove session.
+        // but mqtt protocol version lower than 5.0 we don't do that.
+        if (MqttUtils.isMqtt5(protocolVersion)
+                && SESSION_EXPIRE_INTERVAL_UPDATER.get(this)
+                != SessionExpireInterval.NEVER_EXPIRE.getSecondTime()) {
+            if (sessionExpireInterval == SessionExpireInterval.EXPIRE_IMMEDIATELY.getSecondTime()) {
+                return topicSubscriptionManager.removeAllSubscriptions();
+            } else {
+                manager.newSessionExpireInterval(__ ->
+                        topicSubscriptionManager.removeAllSubscriptions(), clientId, SESSION_EXPIRE_INTERVAL_UPDATER.get(this));
+            }
+        }
+        return CompletableFuture.completedFuture(null);
     }
 
     public boolean assignState(ConnectionState expected, ConnectionState newState) {

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/Connection.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/Connection.java
@@ -153,8 +153,6 @@ public class Connection {
             channel.writeAndFlush(mqttDisconnectionAckMessage);
         }
         this.channel.close();
-        // clean subscriptions
-        topicSubscriptionManager.removeAllConsumer();
         if (cleanSession) {
             return topicSubscriptionManager.removeAllSubscriptions();
         }
@@ -165,11 +163,13 @@ public class Connection {
                 != SessionExpireInterval.NEVER_EXPIRE.getSecondTime()) {
             if (sessionExpireInterval == SessionExpireInterval.EXPIRE_IMMEDIATELY.getSecondTime()) {
                 return topicSubscriptionManager.removeAllSubscriptions();
-            } else {
-                manager.newSessionExpireInterval(__ ->
-                        topicSubscriptionManager.removeAllSubscriptions(), clientId, SESSION_EXPIRE_INTERVAL_UPDATER.get(this));
             }
+            manager.newSessionExpireInterval(__ -> topicSubscriptionManager.removeAllSubscriptions(),
+                    clientId, SESSION_EXPIRE_INTERVAL_UPDATER.get(this));
+            return CompletableFuture.completedFuture(null);
         }
+        // remove all consumer if we don't need to clean session.
+        topicSubscriptionManager.removeAllConsumer();
         return CompletableFuture.completedFuture(null);
     }
 

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/TopicSubscriptionManager.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/TopicSubscriptionManager.java
@@ -2,9 +2,9 @@
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
- * <p>
- * http://www.apache.org/licenses/LICENSE-2.0
- * <p>
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/TopicSubscriptionManager.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/TopicSubscriptionManager.java
@@ -62,6 +62,7 @@ public class TopicSubscriptionManager {
             FutureUtil.failedFuture(e);
         }
         return topic.unsubscribe(subscriberName)
+                .thenAccept(__ -> subscriptionConsumerPair.getLeft().delete())
                 .thenAccept(__ -> topicSubscriptions.remove(topic));
     }
 

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/TopicSubscriptionManager.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/TopicSubscriptionManager.java
@@ -1,0 +1,86 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.streamnative.pulsar.handlers.mqtt;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Maps;
+import io.streamnative.pulsar.handlers.mqtt.exception.MQTTNoSubscriptionExistedException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.pulsar.broker.service.BrokerServiceException;
+import org.apache.pulsar.broker.service.Consumer;
+import org.apache.pulsar.broker.service.Subscription;
+import org.apache.pulsar.broker.service.Topic;
+import org.apache.pulsar.common.util.FutureUtil;
+
+@Slf4j
+public class TopicSubscriptionManager {
+
+    private final Map<Topic, Pair<Subscription, Consumer>> topicSubscriptions = Maps.newConcurrentMap();
+
+    public Pair<Subscription, Consumer> putIfAbsent(Topic topic, Subscription subscription, Consumer consumer) {
+        return topicSubscriptions.put(topic, Pair.of(subscription, consumer));
+    }
+
+    public void removeAllConsumer() {
+        topicSubscriptions.forEach((k, v) -> {
+            try {
+                removeConsumerIfExist(v.getLeft(), v.getRight());
+            } catch (BrokerServiceException ex) {
+                log.warn("subscription [{}] remove consumer {} error",
+                        v.getLeft(), v.getRight(), ex);
+            }
+        });
+    }
+
+    public CompletableFuture<Void> unsubscribe(Topic topic) {
+        Pair<Subscription, Consumer> subscriptionConsumerPair = topicSubscriptions.get(topic);
+        if (subscriptionConsumerPair == null) {
+            return FutureUtil.failedFuture(new MQTTNoSubscriptionExistedException(
+                    String.format("Can not found subscription for topic %s when unSubscribe", topic)));
+        }
+        String subscriberName = subscriptionConsumerPair.getLeft().getName();
+        try {
+            removeConsumerIfExist(subscriptionConsumerPair.getLeft(), subscriptionConsumerPair.getValue());
+        } catch (BrokerServiceException e) {
+            log.error("[ Subscription ] Subscription {} Remove consumer fail.", subscriberName, e);
+            FutureUtil.failedFuture(e);
+        }
+        return topic.unsubscribe(subscriberName)
+                .thenAccept(__ -> topicSubscriptions.remove(topic));
+    }
+
+    public CompletableFuture<Void> removeAllSubscriptions() {
+        List<CompletableFuture<Void>> futures = topicSubscriptions.keySet()
+                .stream()
+                .map(this::unsubscribe)
+                .collect(Collectors.toList());
+        return FutureUtil.waitForAll(futures);
+    }
+
+    @VisibleForTesting
+    public Map<Topic, Pair<Subscription, Consumer>> getTopicSubscriptions() {
+        return topicSubscriptions;
+    }
+
+    private void removeConsumerIfExist(Subscription subscription, Consumer consumer) throws BrokerServiceException {
+        if (subscription.getConsumers().contains(consumer)) {
+            subscription.removeConsumer(consumer);
+        }
+    }
+}

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/TopicSubscriptionManager.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/TopicSubscriptionManager.java
@@ -48,7 +48,7 @@ public class TopicSubscriptionManager {
         });
     }
 
-    public CompletableFuture<Void> unsubscribe(Topic topic) {
+    public CompletableFuture<Void> unsubscribe(Topic topic, boolean cleanSubscription) {
         Pair<Subscription, Consumer> subscriptionConsumerPair = topicSubscriptions.get(topic);
         if (subscriptionConsumerPair == null) {
             return FutureUtil.failedFuture(new MQTTNoSubscriptionExistedException(
@@ -62,14 +62,18 @@ public class TopicSubscriptionManager {
             FutureUtil.failedFuture(e);
         }
         return topic.unsubscribe(subscriberName)
-                .thenAccept(__ -> subscriptionConsumerPair.getLeft().delete())
+                .thenAccept(__ -> {
+                    if (cleanSubscription) {
+                        subscriptionConsumerPair.getLeft().delete();
+                    }
+                })
                 .thenAccept(__ -> topicSubscriptions.remove(topic));
     }
 
     public CompletableFuture<Void> removeAllSubscriptions() {
         List<CompletableFuture<Void>> futures = topicSubscriptions.keySet()
                 .stream()
-                .map(this::unsubscribe)
+                .map(topic -> unsubscribe(topic, true))
                 .collect(Collectors.toList());
         return FutureUtil.waitForAll(futures);
     }

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/proxy/MQTTProxyProtocolMethodProcessor.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/proxy/MQTTProxyProtocolMethodProcessor.java
@@ -141,8 +141,8 @@ public class MQTTProxyProtocolMethodProcessor extends AbstractCommonProtocolMeth
             log.warn("connection is null. close CId={}", clientId);
             channel.close();
         } else {
-            connectionManager.removeConnection(connection);
-            connection.close();
+            connection.close()
+                    .thenAccept(__->connectionManager.removeConnection(connection));
         }
     }
 

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/DefaultProtocolMethodProcessorImpl.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/DefaultProtocolMethodProcessorImpl.java
@@ -435,7 +435,7 @@ public class DefaultProtocolMethodProcessorImpl extends AbstractCommonProtocolMe
                                         String.format("Can not found topic when %s unSubscribe.", clientId));
                             }
                             return connection.getTopicSubscriptionManager()
-                                    .unsubscribe(topicOp.get());
+                                    .unsubscribe(topicOp.get(), false);
                         }))
                         .collect(Collectors.toList());
                 return FutureUtil.waitForAll(futures);

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/DefaultProtocolMethodProcessorImpl.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/DefaultProtocolMethodProcessorImpl.java
@@ -38,7 +38,6 @@ import io.streamnative.pulsar.handlers.mqtt.OutstandingPacket;
 import io.streamnative.pulsar.handlers.mqtt.OutstandingPacketContainer;
 import io.streamnative.pulsar.handlers.mqtt.PacketIdGenerator;
 import io.streamnative.pulsar.handlers.mqtt.QosPublishHandlers;
-import io.streamnative.pulsar.handlers.mqtt.exception.MQTTNoSubscriptionExistedException;
 import io.streamnative.pulsar.handlers.mqtt.exception.MQTTServerException;
 import io.streamnative.pulsar.handlers.mqtt.exception.MQTTTopicNotExistedException;
 import io.streamnative.pulsar.handlers.mqtt.exception.handler.MopExceptionHelper;
@@ -60,20 +59,17 @@ import io.streamnative.pulsar.handlers.mqtt.utils.WillMessage;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
-import java.util.Map;
 import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
 import lombok.extern.slf4j.Slf4j;
 import org.apache.bookkeeper.mledger.impl.PositionImpl;
 import org.apache.commons.lang3.tuple.Pair;
 import org.apache.pulsar.broker.PulsarService;
 import org.apache.pulsar.broker.authentication.AuthenticationDataCommand;
 import org.apache.pulsar.broker.authorization.AuthorizationService;
-import org.apache.pulsar.broker.service.Consumer;
 import org.apache.pulsar.broker.service.Subscription;
-import org.apache.pulsar.broker.service.Topic;
 import org.apache.pulsar.common.api.proto.CommandAck;
 import org.apache.pulsar.common.naming.TopicName;
 import org.apache.pulsar.common.util.FutureUtil;
@@ -92,10 +88,10 @@ public class DefaultProtocolMethodProcessorImpl extends AbstractCommonProtocolMe
     private final AuthorizationService authorizationService;
     private final MQTTMetricsCollector metricsCollector;
     private final MQTTConnectionManager connectionManager;
-    private final MQTTSubscriptionManager subscriptionManager;
+    private final MQTTSubscriptionManager mqttSubscriptionManager;
     private Connection connection;
 
-    public DefaultProtocolMethodProcessorImpl (MQTTService mqttService, ChannelHandlerContext ctx) {
+    public DefaultProtocolMethodProcessorImpl(MQTTService mqttService, ChannelHandlerContext ctx) {
         super(mqttService.getAuthenticationService(),
                 mqttService.getServerConfiguration().isMqttAuthenticationEnabled(), ctx);
         this.pulsarService = mqttService.getPulsarService();
@@ -106,7 +102,7 @@ public class DefaultProtocolMethodProcessorImpl extends AbstractCommonProtocolMe
         this.authorizationService = mqttService.getAuthorizationService();
         this.metricsCollector = mqttService.getMetricsCollector();
         this.connectionManager = mqttService.getConnectionManager();
-        this.subscriptionManager = mqttService.getSubscriptionManager();
+        this.mqttSubscriptionManager = mqttService.getSubscriptionManager();
         this.serverCnx = new MQTTServerCnx(pulsarService, ctx);
     }
 
@@ -251,9 +247,8 @@ public class DefaultProtocolMethodProcessorImpl extends AbstractCommonProtocolMe
             }
         }
         metricsCollector.removeClient(NettyUtils.getAddress(channel));
-        connectionManager.removeConnection(connection);
-        connection.removeSubscriptions();
-        connection.close();
+        connection.close()
+                .thenAccept(__ -> connectionManager.removeConnection(connection));
     }
 
     private boolean checkAndUpdateSessionExpireIntervalIfNeed(String clientId,
@@ -288,13 +283,15 @@ public class DefaultProtocolMethodProcessorImpl extends AbstractCommonProtocolMe
             log.debug("[Connection Lost] [{}] ", clientId);
         }
         metricsCollector.removeClient(NettyUtils.getAddress(channel));
-        connectionManager.removeConnection(connection);
-        connection.removeSubscriptions();
-        subscriptionManager.removeSubscription(clientId);
         WillMessage willMessage = connection.getWillMessage();
         if (willMessage != null) {
             fireWillMessage(willMessage);
         }
+        connection.close()
+                .thenAccept(__ -> {
+                    connectionManager.removeConnection(connection);
+                    mqttSubscriptionManager.removeSubscription(clientId);
+                });
     }
 
     @Override
@@ -303,7 +300,7 @@ public class DefaultProtocolMethodProcessorImpl extends AbstractCommonProtocolMe
     }
 
     private void fireWillMessage(WillMessage willMessage) {
-        List<Pair<String, String>> subscriptions = subscriptionManager.findMatchTopic(willMessage.getTopic());
+        List<Pair<String, String>> subscriptions = mqttSubscriptionManager.findMatchTopic(willMessage.getTopic());
         MqttPublishMessage msg = createMqttWillMessage(willMessage);
         for (Pair<String, String> entry : subscriptions) {
             Connection connection = connectionManager.getConnection(entry.getLeft());
@@ -365,9 +362,8 @@ public class DefaultProtocolMethodProcessorImpl extends AbstractCommonProtocolMe
     private CompletableFuture<Void> doSubscribe(MqttSubscribeMessage msg) {
         final int messageID = msg.variableHeader().messageId();
         final List<MqttTopicSubscription> subTopics = topicSubscriptions(msg);
-        subscriptionManager.addSubscriptions(connection.getClientId(), subTopics);
+        mqttSubscriptionManager.addSubscriptions(connection.getClientId(), subTopics);
         List<CompletableFuture<Void>> futureList = new ArrayList<>(subTopics.size());
-        Map<Topic, Pair<Subscription, Consumer>> topicSubscriptions = new ConcurrentHashMap<>();
         for (MqttTopicSubscription subTopic : subTopics) {
             metricsCollector.addSub(subTopic.topicName());
             CompletableFuture<List<String>> topicListFuture = PulsarTopicUtils.asyncGetTopicListFromTopicSubscription(
@@ -387,7 +383,7 @@ public class DefaultProtocolMethodProcessorImpl extends AbstractCommonProtocolMe
                                     outstandingPacketContainer, metricsCollector, connection.getClientReceiveMaximum());
                             sub.addConsumer(consumer);
                             consumer.addAllPermits();
-                            topicSubscriptions.putIfAbsent(sub.getTopic(), Pair.of(sub, consumer));
+                            connection.getTopicSubscriptionManager().putIfAbsent(sub.getTopic(), sub, consumer);
                         } catch (Exception e) {
                             throw new MQTTServerException(e);
                         }
@@ -408,11 +404,6 @@ public class DefaultProtocolMethodProcessorImpl extends AbstractCommonProtocolMe
                 log.debug("Sending SUB-ACK message {} to {}", ackMessage, connection.getClientId());
             }
             channel.writeAndFlush(ackMessage);
-            Map<Topic, Pair<Subscription, Consumer>> existedSubscriptions = NettyUtils.getTopicSubscriptions(channel);
-            if (existedSubscriptions != null) {
-                topicSubscriptions.putAll(existedSubscriptions);
-            }
-            NettyUtils.setTopicSubscriptions(channel, topicSubscriptions);
         }).exceptionally(e -> {
             log.error("[{}] Failed to process MQTT subscribe.", connection.getClientId(), e);
             MopExceptionHelper.handle(MqttMessageType.SUBSCRIBE, messageID, channel, e);
@@ -427,7 +418,6 @@ public class DefaultProtocolMethodProcessorImpl extends AbstractCommonProtocolMe
             log.debug("[Unsubscribe] [{}] msg: {}", clientId, msg);
         }
         final List<String> topicFilters = msg.payload().topics();
-        final MqttQoS qos = msg.fixedHeader().qosLevel();
         final List<CompletableFuture<Void>> futureList = new ArrayList<>(topicFilters.size());
         for (String topicFilter : topicFilters) {
             metricsCollector.removeSub(topicFilter);
@@ -435,37 +425,19 @@ public class DefaultProtocolMethodProcessorImpl extends AbstractCommonProtocolMe
                     topicFilter, configuration.getDefaultTenant(), configuration.getDefaultNamespace(), pulsarService,
                     configuration.getDefaultTopicDomain());
             CompletableFuture<Void> future = topicListFuture.thenCompose(topics -> {
-                List<CompletableFuture<Void>> futures = new ArrayList<>();
-                for (String topic : topics) {
-                    PulsarTopicUtils.getTopicReference(pulsarService, topic, configuration.getDefaultTenant(),
-                            configuration.getDefaultNamespace(), false,
-                            configuration.getDefaultTopicDomain(), false).thenAccept(topicOp -> {
-                        if (!topicOp.isPresent()) {
-                            throw new MQTTTopicNotExistedException(
-                                    String.format("Can not found topic %s when %s unSubscribe.", topic,
-                                            clientId));
-                        }
-                        Subscription subscription = topicOp.get().getSubscription(clientId);
-                        if (subscription == null) {
-                            throw new MQTTNoSubscriptionExistedException(
-                                    String.format(
-                                            "Can not found subscription for topic %s when %s unSubscribe",
-                                            topic, clientId));
-                        }
-                        try {
-                            MQTTConsumer consumer = new MQTTConsumer(subscription, topicFilter,
-                                    topic, clientId, serverCnx, qos, packetIdGenerator,
-                                    outstandingPacketContainer, metricsCollector, connection.getClientReceiveMaximum());
-                            topicOp.get().getSubscription(clientId).removeConsumer(consumer);
-                            futures.add(topicOp.get().unsubscribe(clientId));
-                        } catch (Exception e) {
-                            throw new MQTTServerException(e);
-                        }
-                    }).exceptionally(ex -> {
-                        futures.add(FutureUtil.failedFuture(ex));
-                        return null;
-                    });
-                }
+                List<CompletableFuture<Void>> futures = topics.stream()
+                        .map(topic -> PulsarTopicUtils.getTopicReference(pulsarService,
+                                topic, configuration.getDefaultTenant(), configuration.getDefaultNamespace(), false,
+                                configuration.getDefaultTopicDomain(), false))
+                        .map(topicFuture -> topicFuture.thenCompose(topicOp -> {
+                            if (!topicOp.isPresent()) {
+                                throw new MQTTTopicNotExistedException(
+                                        String.format("Can not found topic when %s unSubscribe.", clientId));
+                            }
+                            return connection.getTopicSubscriptionManager()
+                                    .unsubscribe(topicOp.get());
+                        }))
+                        .collect(Collectors.toList());
                 return FutureUtil.waitForAll(futures);
             });
             futureList.add(future);

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/utils/NettyUtils.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/utils/NettyUtils.java
@@ -15,16 +15,10 @@ package io.streamnative.pulsar.handlers.mqtt.utils;
 
 import static io.streamnative.pulsar.handlers.mqtt.Constants.ATTR_CLIENT_ADDR;
 import static io.streamnative.pulsar.handlers.mqtt.Constants.ATTR_CONNECTION;
-import static io.streamnative.pulsar.handlers.mqtt.Constants.ATTR_TOPIC_SUBS;
 import io.netty.channel.Channel;
 import io.netty.util.AttributeKey;
 import io.streamnative.pulsar.handlers.mqtt.Connection;
 import java.net.InetSocketAddress;
-import java.util.Map;
-import org.apache.commons.lang3.tuple.Pair;
-import org.apache.pulsar.broker.service.Consumer;
-import org.apache.pulsar.broker.service.Subscription;
-import org.apache.pulsar.broker.service.Topic;
 
 /**
  * Some Netty's channels utilities.
@@ -32,19 +26,10 @@ import org.apache.pulsar.broker.service.Topic;
 public final class NettyUtils {
 
     public static final AttributeKey<Object> ATTR_KEY_CONNECTION = AttributeKey.valueOf(ATTR_CONNECTION);
-    private static final AttributeKey<Object> ATTR_KEY_TOPIC_SUBS = AttributeKey.valueOf(ATTR_TOPIC_SUBS);
     private static final AttributeKey<Object> ATTR_KEY_CLIENT_ADDR = AttributeKey.valueOf(ATTR_CLIENT_ADDR);
 
     public static Connection getConnection(Channel channel) {
         return (Connection) channel.attr(NettyUtils.ATTR_KEY_CONNECTION).get();
-    }
-
-    public static void setTopicSubscriptions(Channel channel, Map<Topic, Pair<Subscription, Consumer>> topicSubs) {
-        channel.attr(NettyUtils.ATTR_KEY_TOPIC_SUBS).set(topicSubs);
-    }
-
-    public static Map<Topic, Pair<Subscription, Consumer>> getTopicSubscriptions(Channel channel) {
-        return (Map<Topic, Pair<Subscription, Consumer>>) channel.attr(NettyUtils.ATTR_KEY_TOPIC_SUBS).get();
     }
 
     public static String getAndSetAddress(Channel channel) {

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/mqtt/mqtt5/hivemq/base/MQTT5CleanStartTest.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/mqtt/mqtt5/hivemq/base/MQTT5CleanStartTest.java
@@ -20,8 +20,10 @@ import io.streamnative.pulsar.handlers.mqtt.base.MQTTTestBase;
 import io.streamnative.pulsar.handlers.mqtt.utils.PulsarTopicUtils;
 import java.util.List;
 import java.util.Objects;
+import java.util.concurrent.TimeUnit;
 import lombok.extern.slf4j.Slf4j;
 import org.apache.commons.collections.CollectionUtils;
+import org.awaitility.Awaitility;
 import org.testng.Assert;
 import org.testng.annotations.Test;
 
@@ -47,8 +49,9 @@ public class MQTT5CleanStartTest extends MQTTTestBase {
         String clientId = Objects.requireNonNull(client.getConfig().getClientIdentifier().orElse(null)).toString();
         Assert.assertTrue(subscriptions.contains(clientId));
         client.disconnect();
-        List<String> afterDisconnectionSubscriptions = admin.topics().getSubscriptions(pulsarTopicName);
-        Assert.assertTrue(CollectionUtils.isEmpty(afterDisconnectionSubscriptions));
+        Awaitility.await()
+                .untilAsserted(() ->
+                        Assert.assertTrue(CollectionUtils.isEmpty(admin.topics().getSubscriptions(pulsarTopicName))));
     }
 
     @Test(timeOut = TIMEOUT)
@@ -71,7 +74,8 @@ public class MQTT5CleanStartTest extends MQTTTestBase {
         String clientId = Objects.requireNonNull(client.getConfig().getClientIdentifier().orElse(null)).toString();
         Assert.assertTrue(subscriptions.contains(clientId));
         client.disconnect();
-        List<String> afterDisconnectionSubscriptions = admin.topics().getSubscriptions(pulsarTopicName);
-        Assert.assertTrue(afterDisconnectionSubscriptions.contains(clientId));
+        Awaitility.await().pollDelay(1, TimeUnit.SECONDS)
+                        .untilAsserted(()->
+                                Assert.assertTrue(admin.topics().getSubscriptions(pulsarTopicName).contains(clientId)));
     }
 }

